tests/src/test/scala/akka/kafka/scaladsl/RebalanceExtSpec.scala

```diff
@@ -0,0 +1,427 @@
+        (0 until partitionCount).toSet.map { partitionIdx: Int =>
+          val startMessageIdx = partitionIdx * perPartitionMessageCount + 1 + topicOffset
+          val endMessageIdx = startMessageIdx + perPartitionMessageCount - 1
+          val messageRange = startMessageIdx to endMessageIdx
+          messageRange.foreach(messageId => {
+            val topicPartition = s"$topic1-$partitionIdx"
+            msgTpMap = msgTpMap.updated(messageId, topicPartition)
+            messageStoreAndAck =
+              messageStoreAndAck.updated(messageId, (new AtomicInteger(0), Promise[Done], Promise[Done]))
+          })
+          produce(topic1, messageRange, partitionIdx).map { f =>
+            val topicPartition = s"$topic1-$partitionIdx"
+            log.debug(s"Produced messages $startMessageIdx to $endMessageIdx to topicPartition $topicPartition")
+            f
+          }
+        }.toSeq
+      }
+      .toSeq
+      .flatten
+    (msgTpMap, producers, messageStoreAndAck)
+  }
+
+  def businessFlow1(clientId: String,
+                    perPartitionMessageCount: Int,
+                    topicPartitionFutureMap: Map[String, Promise[Done]],
+                    messageStoreAndAck: Map[Int, (AtomicInteger, Promise[Done], Promise[Done])],
+                    sharedKillSwitch: SharedKillSwitch,
+                    logPrefix: String): Flow[CommittableMessage[String, String], CommittableOffset, NotUsed] = {
+    Flow.fromFunction { message =>
+      val messageVal = message.record.value.toInt
+      val duplicateCount = messageStoreAndAck(messageVal)._1.incrementAndGet()
+      val topicPartition = s"${message.record.topic}-${message.record.partition}"
+      val msg1 =
+        s"$logPrefix::businessFlow1:offset=${message.committableOffset.partitionOffset.offset} messageId=$messageVal topicPartition=${message.record.topic}-${message.record.partition} consumerId=$clientId duplicateCount=$duplicateCount"
+      if (duplicateCount > 1) {
+        log.warn(s"businessFlow1:received:$msg1")
+      } else {
+        log.info(s"businessFlow1:received:$msg1")
+      }
+      // by default threads are blocked here, waiting for a manual unblock as per testing steps
+      log.info(s"$logPrefix::businessFlow1::BEGIN:blockAtMessage=$messageVal:: $msg1")
+      Await.result(messageStoreAndAck(messageVal)._2.future, maxAwait)
+      val promise1 = messageStoreAndAck(messageVal)._3
+      if (!promise1.isCompleted) {
+        promise1.complete(Try(Done))
+        log.info(s"$logPrefix::businessFlow1:DONE:blockAtMessage=$messageVal:: $msg1")
+      } else {
+        log.info(s"$logPrefix::businessFlow1:IGNORE:blockAtMessage=$messageVal:: $msg1")
+      }
+      log.info(s"$logPrefix::businessFlow1::END:blockAtMessage=$messageVal:: $msg1")
+      if (message.committableOffset.partitionOffset.offset == perPartitionMessageCount - 1) {
+        val promise1 = topicPartitionFutureMap(topicPartition)
```

Comment:
Maybe give this promise a more meaningful name, same with the `promise1` one scope up ^

Reply:
done